### ISSUE TITLE
Add Amber ff19SB params

### ISSUE THREAD

**Issue by aizvorski** [openmm/openmm#3655]
Reference: https://pubs.acs.org/doi/10.1021/acs.jctc.9b00591 or https://chemrxiv.org/engage/chemrxiv/article-details/60c7426b0f50db0f23395cd7

From the Amber22 manual, "ff19SB [19] is the latest model of the SB protein forcefields, developed in the Simmerling Lab at Stony Brook University. The new ff19SB forcefield has shown to improve amino acid-dependent properties such as helical propensities and reproduces the differences in amino-acid-specific PDB Ramachandran map.  Users are encouraged to read the ff19SB article [19] to learn more about the motivation behind ff19SB, as well as details of the fitting and testing protocols and improved performance relative to ff14SB. Our older SB protein forcefield models utilized uncoupled phi/psi dihedral parameters for the protein backbone, and every amino acid except for glycine used the backbone dihedral parameters fit using alanine. In ff19SB, we improved the backbone dihedrals parameters for every standard amino acids. We fit coupled φ/ψ parameters using 2D φ/ψ conformational scans for multiple amino acids, using 2D QM energy surfaces in solution as reference data. These new dihedral parameters include amino-acid specific CMAPs that are based on residue name. "

I'd be happy to help do the conversion or test this; I assume it would mostly involve running `openmm/devtools/forcefield-scripts/processAmberForceField.py` but I haven't ran that before so I am not sure how to proceed. Recent versions of ambertools have `ambertools/dat/leap/parm/frcmod.ff19SB` and so forth.

**Comment by aizvorski** [comment]
By the way, this seems to succeed and outputs a reasonable xml file:

`python openmm/devtools/forcefield-scripts/processAmberForceField.py ambertools-20.9/dat/leap/parm/frcmod.ff19SB`

However the frcmode file has a bunch of CMAP sections, and it seems they are just ignored (?)

**Comment by aizvorski** [comment]
@peastman From reading the parmed issue, these CMAPs are per residue not per atom type?  So if we created copies of each atom type for each residue we could have a xml that assigns them in the normal way?

**Comment by peastman** [comment]
Yes, they could have chosen to define them that way.  And in principle one could programmatically synthesize new atom types to represent them.  That would take big changes to ParmEd though.

Atom types are only used during parametrization.  That's why we don't have any problem with loading a prmtop that uses ff19.  The parameter assignment has already been done.

**Comment by jchodera** [comment]
@aizvorski : We'd be delighted if someone wanted to put in the effort to modify ParmEd to enable the conversion of ff19SB, but we're not able to do the work here, and effort channeled into more modern workflows (such as the [Open Force Field Initiative](http://openforcefield.org) approach of using SMARTS matches to apply parameters) is likely to pay much higher dividends over the long term.

**Comment by aizvorski** [comment]
@jchodera That makes sense!  However it looks like the SMIRNOFF spec doesn't have CMAPs yet https://github.com/openforcefield/standards/issues/14

There's basically three directions I can see for getting this working:
1. hack in per-residue (rather than per-atom) CMAPs in openmm ForceField
2. add CMAPs to SMIRNOFF, and add ability to parametrize proteins using a SMIRNOFF file to openmm (possibly on top of openff-toolkit rather than new code)
3. hack in a whole bunch of extra atom types (every residue's backbone atoms become a new type) and create an openmm forcefield xml file which does per-atom-type CMAPs for ff19SB, based on the extra atom types
(possibly do some stuff in parmed - I wasn't really considering this option though)

Of these, 1 requires probably a bunch of new code in an area of code I'm not very familiar with, and is also not at all reusable; 2 is very reusable as you said, but requires a lot of foundational work on the openff side.  3 seemed like the path of least resistance, as it can be done with zero new code - it is inelegant, but also (in principle) not hard.

@peastman Do openmm forcefield xml files support CMAPs right now?  (okay if they're per atom type not per residue)  Just want to confirm that the "duplicate atom types" method of doing this is feasible in principle.

**Comment by jchodera** [comment]
> That makes sense! However it looks like the SMIRNOFF spec doesn't have CMAPs yet

That's right---integrating it will be part of a process whereby its value in delivering increased accuracy on a large set of benchmark tasks is demonstrated when fit to the same data. This is an experiment that has not yet been carried out.

> @peastman Do openmm forcefield xml files support CMAPs right now? (okay if they're per atom type not per residue) Just want to confirm that the "duplicate atom types" method of doing this is feasible in principle.

Yes, the [`openmm.app.ForceField` class supports CMAPs](https://github.com/openmm/openmm/blob/master/wrappers/python/openmm/app/forcefield.py#L2325-L2417).

Another possibility would be to add a `<script>...</script>` section to the end of the ffxml file that would iterate through the `Topology` object and add any CMAP terms directly to the appropriate atoms. You could even use the existing `< CMAPTorsionForce>` tag to create the `CMAPTorsionForce` object and add all the maps you will reference, and just have the logic in the `<script>` section for the `force.addTorsion(...)` calls that add the appropriate atoms for the maps.

**Comment by merkys** [comment]
Has there been any progress in AMBER ff19SB support direction?

**Comment by peastman** [comment]
Not that I know of.

**Comment by chuantian** [comment]
Not sure if this comment adds any values (maybe not): CMAP is a 2D function of 5 backbone atoms (C-N-CA-C-N), or, of 2 torsion angles (phi and psi). This new relationship between energy and 2 torsions, 'has' to be defined first. And if it has to be done in a manner of atom type, perhaps it's good to introduce one for every residue (maybe using Calpha) that has associated CMAP. Then, different residues would have "same" Fourier based torsion parameter that's in 1D for phi or psi (that's how it was in ff prior to ff19SB); and "different" CMAP based torsion parameter that's in 2D for phi and psi.

**Comment by epretti** [comment]
ff19SB was added in #4916.
